ft: S3C-3178 bucket notification extension

naren-rajendran opened the pull request:
New extension for bucket notification. Follows the existing extension model.

Commit:
ft: S3C-3178 bucket notification extension

bert-e commented:
# Hello naren-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/backbeat/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/S3C-3178-bucket-notification-extension` into
`development/7.8`
* [w/8.1/feature/S3C-3178-bucket-notification-extension](https://github.com/scality/backbeat/compare/development/8.1...w/8.1/feature/S3C-3178-bucket-notification-extension) will be merged into `development/8.1` (pull request #926)
* [w/8.2/feature/S3C-3178-bucket-notification-extension](https://github.com/scality/backbeat/compare/development/8.2...w/8.2/feature/S3C-3178-bucket-notification-extension) will be merged into `development/8.2` (pull request #927)



The following branches will **NOT** be impacted:

* `development/7.4`
* `development/7.7`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers





Peer approvals *must* include a mandatory approval from @jonathan-gramain.

jonathan-gramain commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -0,0 +1,35 @@
+            return undefined;
+        }
+        if (!isMasterKey(entry.key)) {
+            return this._filterVersionedKey(entry);
```
This function does not look to exist

naren-rajendran replied:
Work in progress :)

bert-e commented:
# History mismatch



Merge commit #8d3ae47099e6d6c8f0f1a513f1f29caa72cf7e59 on the integration branch
`w/8.1/feature/S3C-3178-bucket-notification-extension` is merging a branch which is neither the current
branch `feature/S3C-3178-bucket-notification-extension` nor the development branch
`development/8.1`.

It is likely due to a rebase of the branch `feature/S3C-3178-bucket-notification-extension` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Commit:
ft: S3C-3178 zookeeper node name correction

miniscruff commented on extensions/notification/NotificationConfigValidator.js:
```diff
@@ -0,0 +1,21 @@
+const joi = require('@hapi/joi');
+
+const joiSchema = {
```
joiSchema is a little vague, maybe saying what its the schema for? Or something similar?

naren-rajendran replied:
Handling tests separately.  `joiSchema` I was following other extensions in backbeat as reference, so kept the same name. Will modify it if it makes sense.

miniscruff replied:
Are you handling tests in another PR? Why do tests separately?

naren-rajendran replied:
Separate commit, still figuring it out!

miniscruff replied:
Might I introduce you to Martin Fowler and TDD https://martinfowler.com/bliki/TestDrivenDevelopment.html

naren-rajendran replied:
This is still a work in progress, when I meant figuring things out I meant there are some gaps I need to address. I will add the tests once done.

naren-rajendran replied:
I accidentally knocked the label WIP off.

miniscruff replied:
Oh ok.

miniscruff reviewed (commented):
Small name thing, but also no tests.

naren-rajendran commented:
Accidentally removed WIP label.

dora-korpar commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -0,0 +1,220 @@
+        const zkPath = this._getBucketNodeZkPath(bucket);
+        return this.zkClient.getData(zkPath, (err, data) => {
+            if (err && err.name !== 'NO_NODE') {
+                this.log.error('Could not fetch bucket notif config', {
```
teeeeeny tiny, but it might be useful to write out `notification configuration` in this error

dora-korpar commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -0,0 +1,220 @@
+                return done(err);
+            }
+            if (!err) {
+                this.log.info('removed bucket notif config zookeeper node', {
```
same with this log

dora-korpar reviewed (approved):
lgtm

Commit:
ft: S3C-3178 updates in processor

- constructor param changes in queue processor and kafka destination
- updated config filter implemented
- new constants added for notification

Commit:
ft: S3C-3178 log message corrections

Commit:
ft: S3C-3178 config support update

Commit:
ft: S3C-3178 linter corrections

Commit:
ft: 3178 support auth for kafka client

jonathan-gramain commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -0,0 +1,224 @@
+    require('../../lib/queuePopulator/QueuePopulatorExtension');
+
+class NotificationQueuePopulator extends QueuePopulatorExtension {
+    constructor(params) {
```
Might be useful to provide jsdoc to the constructor to know what's expected in `params`

jonathan-gramain commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -0,0 +1,224 @@
+                const { error, result } = safeJsonParse(data);
+                if (error) {
+                    this.log.error('invalid config', { method, zkPath, data });
+                    return done(null, 1);
```
Not fond of returning special numeric values for special cases, why not using specific error types for this? say `ObjNotFound` or `InvalidXyz` or `BadXyz`, or a custom error object created locally that you can compare to, etc., in all those cases no value is returned, so it can be considered an error, even if some errors are "normal".

naren-rajendran replied:
I will address this in an improvement task.

jonathan-gramain commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -0,0 +1,224 @@
+        });
+    }
+
+    filter(entry) {
```
From a style perspective, this function looks a bit monolithic with too many indentation levels. Ideally we would have split the logic into separate methods.

As a side note, I'm not a fan of the limit of 80 characters per line (going to repeat this every time :smiling_imp: ), but at least a benefit is it should discourage to have too many nested blocks.

naren-rajendran replied:
I will address this in an improvement task.

jonathan-gramain commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -0,0 +1,224 @@
+        }
+        // bucket notification configuration updates
+        if (key === undefined) {
+            if (bucket && result) {
```
Recommend to put all this block for bucket updates (`key === undefined`) in a separate method `filterBucketUpdate` for example, for readability, and then can be split further into `filterBucketPut` and `filterBucketDelete`.

naren-rajendran replied:
I will address this in an improvement task.

jonathan-gramain commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -0,0 +1,224 @@
+        if (key && result) {
+            // TODO: handle versioned object entry
+            return async.waterfall([
+                next => this._getBucketNotifConfig(bucket, next),
```
I see a potential for race condition here and in general in this function, because we're doing asynchronous processing while the function is synchronous and have no flow control on the async tasks it triggers.

For example, when a new bucket notification configuration is discovered, it will be written asynchronously to zookeeper, but in the meantime the new object entries that are fed to `filter` will not benefit from this configuration until the configuration is effectively written, although such object entries would have been added after the bucket notification configuration was set on the bucket.

It may not be a major issue for now, but we should eventually revisit this I think. One way could be to implement the local cache of configurations, so that we could apply them immediately as they are discovered.

naren-rajendran replied:
Yes, we had this discussion before in one of our sessions, and caching was discussed as improvement. Definitely this has to be addressed in the next iteration.

jonathan-gramain commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -0,0 +1,224 @@
+                                bucket,
+                                JSON.stringify(ent));
+                        }
+                        return next();
```
minor: this call can be removed as it falls back to the next one

jonathan-gramain commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -0,0 +1,224 @@
+            return async.waterfall([
+                next => this._getBucketNotifConfig(bucket, next),
+                (config, next) => {
+                    if (config && Object.keys(config).length > 0) {
```
Not a real issue, but technically this section has no asynchronous part, so it could be more proper to do it in the callback of the `async.waterfall`

jonathan-gramain commented on extensions/notification/constants.js:
```diff
@@ -1,8 +1,20 @@
+    zkBucketNotificationPath: '/bucket-notification',
+    bucketNotifConfigPropName: 'notificationConfiguration',
+    notificationEventPropName: 'originOp',
+    eventTimePropName: 'last-modified',
```
Wondering if it's not better to just use `last-modified` than have a specific constant for this, as constants are normally most useful for unusual values or to avoid repeating or typo-ing them, but reading `last-modified` makes sense for someone accustomed to HTTP and is self-explanatory.

naren-rajendran replied:
I just added it to constants, if we decide to use some other field, but going to keep this for the time being.

jonathan-gramain commented on extensions/notification/queueProcessor/QueueProcessor.js:
```diff
@@ -24,31 +30,89 @@ class QueueProcessor extends EventEmitter {
+        const zookeeperUrl =
+            `${this.zkConfig.connectionString}${populatorZkPath}`;
+        this.logger.info('opening zookeeper connection for persisting ' +
+            'populator state',
```
This is to read the state, not to persist it

jonathan-gramain commented on extensions/notification/utils/auth.js:
```diff
@@ -0,0 +1,72 @@
+            fs.accessSync(path, fs.F_OK | fs.R_OK);
+            return path;
+        } catch (e) {
+            return '';
```
Why not `return null` instead to more clearly make it an error.

jonathan-gramain commented on extensions/notification/utils/auth.js:
```diff
@@ -0,0 +1,72 @@
+        }
+        authObject['ssl.key.password'] = keyPassword;
+    }
+    // only kereberos is supported now!
```
typo: ker ~e~ beros, remove the "!"

jonathan-gramain reviewed (approved):
Mostly minor comments, not expecting to see them all fixed in the PR but can serve as guidelines for future changes.

Commit:
ft: S3C-3178 Minor updates and corrections

- Wildcard support for event filters
- Wildcard support unit tests updated
- Minor refactoring to NotificationQueuePopulator.

Commit:
ft: S3C-3178 bucket notification updates

- Removed TODOs that are supposed to be handled by federation like
  extension config management and zookeeper path
- Changed 'type' to 'eventType' is the message produced by notification
  corrected necessary tests and util
- Metrics consumer and producer are disabled for notification populator
- Minor refactoring in NotificationPopulator, corrected delete entry
  handling for non versioned objects
- LogReader.js CRR checks disabled if notification extension is set

naren-rajendran commented:
@bert-e approve

bert-e commented:
# History mismatch



Merge commit #8d3ae47099e6d6c8f0f1a513f1f29caa72cf7e59 on the integration branch
`w/8.1/feature/S3C-3178-bucket-notification-extension` is merging a branch which is neither the current
branch `feature/S3C-3178-bucket-notification-extension` nor the development branch
`development/8.1`.

It is likely due to a rebase of the branch `feature/S3C-3178-bucket-notification-extension` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

naren-rajendran commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.1/feature/S3C-3178-bucket-notification-extension` with contents from `feature/S3C-3178-bucket-notification-extension`
and `development/8.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.1/feature/S3C-3178-bucket-notification-extension origin/development/8.1
 $ git merge origin/feature/S3C-3178-bucket-notification-extension
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.1/feature/S3C-3178-bucket-notification-extension
```





*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/S3C-3178-bucket-notification-extension` into
`development/7.8`
* [w/8.1/feature/S3C-3178-bucket-notification-extension](https://github.com/scality/backbeat/compare/development/8.1...w/8.1/feature/S3C-3178-bucket-notification-extension) will be merged into `development/8.1` (pull request #995)
* [w/8.2/feature/S3C-3178-bucket-notification-extension](https://github.com/scality/backbeat/compare/development/8.2...w/8.2/feature/S3C-3178-bucket-notification-extension) will be merged into `development/8.2` (pull request #996)



The following branches will **NOT** be impacted:

* `development/7.4`
* `stabilization/7.4.7`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/backbeat/commit/b368abf7a7a6d69eba132d1590be21a00f8844de) did not succeed in branch w/8.1/feature/S3C-3178-bucket-notification-extension.




*The following options are set:* **approve**

naren-scality commented:
@bert-e reset

naren-scality commented:
@bert-e force_reset

naren-scality commented:
@bert-e help

bert-e commented:
# Help page



The following options and commands are available at this time.

## Options ##


name   | description  | privileged   | authored
------ | ------------ | ------------ |----------
:arrow_right: **no_octopus** | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
:arrow_right: **bypass_peer_approval** | Bypass the pull request peers' approval |  :star:  | 
:arrow_right: **create_pull_requests** | Allow the creation of integration pull requests. |  | 
:arrow_right: **bypass_author_approval** | Bypass the pull request author's approval |  :star:  | 
:arrow_right: **bypass_incompatible_branch** | Bypass the check on the source branch prefix |  :star:  | 
:arrow_right: **bypass_leader_approval** | Bypass the pull request leaders' approval |  :star:  | 
:arrow_right: **unanimity** | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
:arrow_right: **after_pull_request** | Wait for the given pull request id to be merged before continuing with the current one. |  | 
:arrow_right: **bypass_build_status** | Bypass the build and test status |  :star:  | 
:arrow_right: **wait** | Instruct Bert-E not to run until further notice. |  | 
:arrow_right: **bypass_jira_check** | Bypass the Jira issue check |  :star:  | 
:arrow_right: **bypass_commit_size** | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
:arrow_right: **approve** | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 



## Commands ##


name   | description  | privileged
------ | ------------ | ------------
:arrow_right: **clear** | Remove all comments from Bert-E from the history ```TBA``` | 
:arrow_right: **help** | Print Bert-E's manual in the pull request. | 
:arrow_right: **reset** | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
:arrow_right: **status** | Print Bert-E's current status in the pull request ```TBA``` | 
:arrow_right: **force_reset** | Delete integration branches & pull requests, and restart merge process from the beginning. | 
:arrow_right: **build** | Re-start a fresh build ```TBA``` | 
:arrow_right: **retry** | Re-start a fresh build ```TBA``` | 






*The following options are set:* **approve**

naren-scality commented:
@bert-e build

bert-e commented:
# Not implemented



This command will be implemented soon.




*The following options are set:* **approve**

naren-scality commented:
@bert-e retry

bert-e commented:
# Not implemented



This command will be implemented soon.




*The following options are set:* **approve**

naren-scality commented:
@bert-e force_reset

naren-scality commented:
@bert-e force_reset

naren-scality commented:
ping

naren-rajendran commented:
@bert-e approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/7.8`

* :heavy_check_mark: `development/8.1`

* :heavy_check_mark: `development/8.2`



The following branches will **NOT** be impacted:

* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/7.8`

* :heavy_check_mark: `development/8.1`

* :heavy_check_mark: `development/8.2`



The following branches have **NOT** changed:

* `development/7.4`



Please check the status of the associated issue S3C-3178.

Goodbye naren-scality.
